src/cache.c

```diff
@@ -115,76 +115,68 @@ int metal_icache_l1_available(int hartid) {
+ * immediate 12bs)
+ * 31     28 27    24 23    20 19    16 15    12 11     8 7      4 3       0
  * |--------|--------|--------|--------|--------|--------|--------|--------|
  * +-------------+------------+----------+------+--------+-----------------+
  * |sign immediate12b (simm12)|   rs1    | func3|    rd  |      opcode     |
  * |-1-1-1-1 -1-1-0-0 -0-0-0-0|-x-x-x-x-x|0-0-0-|-0-0-0-0|-0-1-1-1 -0-0-1-1|
  * +--------------------------+----------+------+--------+-----------------+
- * 31     -0x3E              20          15  0  12   x0  7      0x73       0
+ * 31     -0x3E              20        15   0 12    x0    7      0x73      0
  * +--------+--------+--------+----------+------+--------+--------+--------+
  * where,
- * rs1 = 0x0, CDISCARD.D.L1 invalidates all lines in the L1 D$ with no writes
+ * rs1 = x0, CDISCARD.D.L1 invalidates all lines in the L1 D$ with no writes
  * back. rs1 != x0, CDISCARD.D.L1 invalidates the L1 D$ line containing the
  * virtual address in integer register rs1, with no writes back.
  */
 void metal_dcache_l1_discard(int hartid, uintptr_t address) {
     if (metal_dcache_l1_available(hartid)) {
-        // Using ‘.insn’ pseudo directive: '.insn i opcode, func3, rd, rs1,
-        // simm12'
-        __asm__ __volatile__(".insn i 0x73, 0, x0, %0, -0x3E" : : "r"(address));
-        __asm__ __volatile__("fence.i"); // FENCE
-    }
-}
-
-/*!
- * @brief CFlush.I.L1 instruction is a custom instruction implemented as a state
- * machine in L1 Instruction Cache (I$) with funct3=0, (for core with data
- * caches) It is an I type: .insn i opcode, func3, rd, rs1, simm12(signed
- * immediate 12bs) 31     28 27    24 23    20 19     16 15   12 11     8 7 4 3
- * 0
- * |--------|--------|--------|--------|--------|--------|--------|--------|
- * +-------------+------------+----------+------+--------+-----------------+
- * |sign immediate12b (simm12)|   rs1    | func3|    rd  |      opcode     |
- * |-1-1-1-1 -1-1-0-0 -0-0-0-0|-0-0-0-0-0|0-0-0-|-0-0-0-0|-0-1-1-1 -0-0-1-1|
- * +--------------------------+----------+------+--------+-----------------+
- * 31     -0x3F              20          15  0  12   x0  7      0x73       0
- * +--------+--------+--------+----------+------+--------+--------+--------+
- * CFLUSH.I.L1 invalidates all lines in the L1 I$.
- */
-void metal_icache_l1_flush(int hartid) {
-    if (metal_icache_l1_available(hartid)) {
-        // Using ‘.insn’ pseudo directive: '.insn i opcode, func3, rd, rs1,
-        // simm12'
-        __asm__ __volatile__(".insn i 0x73, 0, x0, x0, -0x3F" : :);
-        __asm__ __volatile__("fence.i"); // FENCE
+        if (address) {
+            // Using ‘.insn’ pseudo directive:
+            //       '.insn i opcode, func3, rd, rs1, simm12'
+            __asm__ __volatile__(".insn i 0x73, 0, x0, %0, -0x3E"
+                                 :
+                                 : "r"(address));
+        } else {
+            __asm__ __volatile__(".word 0xfc200073" : : : "memory");
```

Comment:
0xfc200073 is not equivalent to .insn i 0x73, 0, x0, %0, -0x3E when %0 = 0, the value should be 0xfc100073